services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientValidator.java

```diff
@@ -116,6 +116,13 @@ public void readJws() throws JWSInputException {
 
         jws = new JWSInput(clientAssertion);
         token = jws.readJsonContent(JsonWebToken.class);
+
+        var event = context.getEvent();
+        if (token != null && event != null) {
```

Comment:
The null checks for token and event seems to be redundant. Did you try it without it?